components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt

```diff
@@ -0,0 +1,413 @@
+    }
+
+    override suspend fun removeAll() {
+        onDiskStorage.removeAll()
+        removeGeckoAllContentPermissions()
+    }
+
+    override suspend fun all(): List<SitePermissions> {
+        val onDiskPermissions: List<SitePermissions> = onDiskStorage.all()
+        val geckoPermissionsByHost = findAllGeckoContentPermissions().groupByDomain()
+
+        return onDiskPermissions.mapNotNull { onDiskPermission ->
+            val map = geckoPermissionsByHost[onDiskPermission.origin].groupByType()
+            mergePermissions(onDiskPermission, map)
+        }
+    }
+
+    @VisibleForTesting
+    internal suspend fun findAllGeckoContentPermissions(): List<ContentPermission>? {
+        return withContext(mainScope.coroutineContext) {
+            geckoStorage.allPermissions.await()
+                .filterNotTemporaryPermissions(geckoTemporaryPermissions)
+        }
+    }
+
+    /**
+     * Updates the [geckoStorage] if the provided [userSitePermissions]
+     * is exists on the [geckoStorage] or it's provided as a part of the [permissionRequest]
+     * otherwise nothing is updated.
+     * @param userSitePermissions the values provided by the user to be updated.
+     * @param permissionRequest the [PermissionRequest] from the web content.
+     * @return An updated [SitePermissions] with default values, if they were updated
+     * on the [geckoStorage] otherwise the same [SitePermissions].
+     */
+    @VisibleForTesting
+    @Suppress("LongMethod")
+    internal suspend fun updateGeckoPermissionIfNeeded(
+        userSitePermissions: SitePermissions,
+        permissionRequest: PermissionRequest? = null
+    ): SitePermissions {
+        var updatedPermission = userSitePermissions
+        val geckoPermissionsByType =
+            permissionRequest.extractGeckoPermissionsOrQueryTheStore(userSitePermissions.origin)
+
+        if (geckoPermissionsByType.isNotEmpty()) {
+            val geckoNotification = geckoPermissionsByType[PERMISSION_DESKTOP_NOTIFICATION]?.firstOrNull()
+            val geckoLocation = geckoPermissionsByType[PERMISSION_GEOLOCATION]?.firstOrNull()
+            val geckoMedia = geckoPermissionsByType[PERMISSION_MEDIA_KEY_SYSTEM_ACCESS]?.firstOrNull()
+            val geckoLocalStorage = geckoPermissionsByType[PERMISSION_PERSISTENT_STORAGE]?.firstOrNull()
+            val geckoAudible = geckoPermissionsByType[PERMISSION_AUTOPLAY_AUDIBLE]?.firstOrNull()
+            val geckoInAudible = geckoPermissionsByType[PERMISSION_AUTOPLAY_INAUDIBLE]?.firstOrNull()
+
+            if (geckoNotification != null) {
+                removeTemporaryPermissionIfAny(geckoNotification)
+                geckoStorage.setPermission(
+                    geckoNotification,
+                    userSitePermissions.notification.toGeckoStatus()
+                )
+                updatedPermission = updatedPermission.copy(notification = NO_DECISION)
+                emitPermissionMigratedFact("notification permission migrated")
```

Comment:
@pocmo I remembered you mentioned the possibility of add some telemetry for when permissions get migrated? As permissions get migrated little by little, do you think is there any additional that we should capture?

Reply:
Let's make sure to keep actually sending the telemetry in Fenix separate though so we don't have to wait for data review.